File: 0316_remove_dupliate_letters/remove_duplicate_letters.py
def removeDuplicateLetters(s: str) -> str:
    """
    Time: O(n)
    Space: O(1)
    """
    seen = set()
    last = {c: i for i, c in enumerate(s)}
    stack = []
    for i, c in enumerate(s):
        if c not in seen:
            while stack and c < stack[-1] and last.get(stack[-1]) > i:
                seen.remove(stack.pop())
            seen.add(c)
            stack.append(c)
    res = ""
    for c in stack:
        res += c
    return res

File: 0316_remove_dupliate_letters/remove_duplicate_letters.java
import java.util.*;

class Solution {
    public String removeDuplicateLetters(String s) {
        Set<Character> seen = new HashSet<Character>();
        Map<Character, Integer> last = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            last.put(s.charAt(i), i);
        }
        Stack<Character> stack = new Stack();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!seen.contains(c)) {
                while (!stack.isEmpty() && c < stack.peek() && i < last.get(stack.peek())) {
                    seen.remove(stack.pop());
                }
                seen.add(c);
                stack.push(c);
            }
        }
        StringBuilder sb = new StringBuilder();
        for (Character c : stack) {
            sb.append(c.charValue());
        }
        return sb.toString();
    }
}